File: ex/hexagonal/src/main/kotlin/com/ktb/shop/domain/port/CustomerRepository.kt
package com.ktb.shop.domain.port

import com.ktb.shop.domain.Customer
import com.ktb.shop.domain.Email

// DRIVEN port: the core calls outward through this contract.
interface CustomerRepository {
    fun findByEmail(email: Email): Customer?
    fun save(customer: Customer)
}

File: ex/hexagonal-go/internal/domain/port/customer_repository.go
package port

import (
	"context"

	"github.com/ktb/hexagonal-shop/internal/domain"
)

// CustomerRepository is a DRIVEN port: the core calls outward through this contract.
// A nil *Customer with a nil error means "not found".
type CustomerRepository interface {
	FindByEmail(ctx context.Context, email domain.Email) (*domain.Customer, error)
	Save(ctx context.Context, customer domain.Customer) error
}
